### ISSUE TITLE
Feature: `maxAttempts` and fail-fast when `stopOnFailure===false`

### ISSUE THREAD

**Issue by adamhamlin** [AlonMiz/poll-until-promise#27]
Hi @AlonMiz,

Curious if you'd be interested in a PR for the following:

- `maxAttempts` -- Sometimes, the semantics of "attempt this operation X times" are preferable over "attempt this operation until X time has passed". You could get _close_ to this behavior by setting `timeout = interval * numDesiredAttempts`, but it's not exact/guaranteed and it also gets more complicated if you're using a `backoffFactor` or similar. Note that this option is additive and wouldn't _replace_ the `timeout`; if `maxAttempts` is passed, then polling stops when either the timeout is reached OR attempts are exhausted.

- `fail-fast when stopOnFailure===false` -- The `stopOnFailure` setting is all-or-nothing; that is, either ANY error will stop the loop, or no errors will. It would be nice to have some dedicated error that could be conditioned to stop the loop even when `stopOnFailure` is false.

```ts
// Usage in execution function:
throw new AbortError(myError); // <--- stops polling immediately, myError is reported
```

**Comment by AlonMiz** [comment]
Hi, thanks for the great suggestions
1. `maxAttempts` seems very reasonable. i agree with the statements. i just wonder about how it will interact with the rest of the properties. eg. if you set both `maxAttempts` and `timeout`, I guess that the first wins.
2. fail-fast seems less intuitive. i guess it depends on the API/Usage. i think that some libraries are using this convenient mechanism:
```ts
await waitFor((abort) => {
  if (something > 0) {
     return abort()
  }
  throw new Error('just retry')
}, {})
```
WDYT?

**Comment by adamhamlin** [comment]
Thanks for replying!

>1. maxAttempts seems very reasonable. i agree with the statements. i just wonder about how it will interact with the rest of the properties. eg. if you set both maxAttempts and timeout, I guess that the first wins.

Yes, agree, first one should win. Definitely still want time bounding as a fallback.

>2. fail-fast seems less intuitive. i guess it depends on the API/Usage. i think that some libraries are using this convenient mechanism:
 
Yes, I did see that `abort` style in some other libs. I think my biggest concern is that it opens up the possibility of bugs/weird behavior because you only wrote `abort()` when you should've written `return abort()`. Kind of a "gotcha".

I stole the `AbortError` idea from [p-retry](https://github.com/sindresorhus/p-retry), which is a widely-used retry library. I find it more clear and user-proof.

I'll go ahead and open a PR and maybe we can hash it out there.

**Comment by AlonMiz** [comment]
Thanks to @adamhamlin,
it's been released in 4.2.0
https://github.com/AlonMiz/poll-until-promise/pull/28

**Comment by adamhamlin** [comment]
Thanks @AlonMiz!
